### PR TITLE
Support leader election for need retry

### PR DESCRIPTION
**Pull request by li-zheng-hao**
to solve this problem ： [https://github.com/dotnetcore/CAP/issues/1271](https://github.com/dotnetcore/CAP/issues/1271)

I have seen the solution in this pr：[https://github.com/dotnetcore/CAP/pull/671](https://github.com/dotnetcore/CAP/pull/671)，It needs to use the third-party component consul 

my solution only use the existing storage module in CAP , and it is simpler

test cases for related functions have been supplemented

@yang-xiaodong  Hi~ do you think this solution is feasible?

### CODE REVIEW

**Comment by li-zheng-hao** [comment]
以防表达不清晰我还是用中文描述：

你好，我刚看了一下[https://medium.com/@bb8s/design-distributed-lock-with-mysql-9bc28ac59629](https://medium.com/@bb8s/design-distributed-lock-with-mysql-9bc28ac59629)这篇文章，其中提到了几点`Functional requirements`：

1. Mutually exclusive.（满足，已通过单元测试）
2. TTL or lease mechanism. （满足TTL，释放机制未实现，这里说明一下： 数据库表中的`RefereshTime`用于控制锁的持有时间，`RetryPrivilegeCheckProcessor`用于定时延长锁持有时间防止死锁及锁被抢夺）
3. High performance (表中只有一条数据，刷新锁持有权频率不算高，个人认为性能不低，满足)
4. Fairness （满足）
5. Re-entrance lock（部分满足，由于实例的`uniqueid`默认在启动时设置一个guid，考虑进程crash后重启该id会刷新，因此会存在一段时间无法重新获取锁，此处应该在进程退出时加上释放锁的逻辑）

此外关于`timestamp`+`unique key `的机制我认为我当前的解决方案中是一致的，表中的`Id`主键为唯一key(固定为0)，`LeaderId`存储当前获取到锁的实例id，`RefreshTime`即`timestamp`，用于防止死锁

总结一下，我认为我的解决方案和该文章的实现机制应该是一致的(如果我没理解错的话~)，不过还有一些地方需要进一步修改

In case the expression is not clear, I will describe it in Chinese:
Hi, I just took a look at [https://medium.com/@bb8s/design-distributed-lock-with-mysql-9bc28ac59629](https://medium.com/@bb8s/design-distributed-lock- with-mysql-9bc28ac59629) article, which mentioned a few `Functional requirements`:

1. Mutually exclusive. (Satisfied, passed unit test)
2. TTL or lease mechanism. (TTL is satisfied, the release mechanism is not implemented, here is an explanation: `RefereshTime` in the database table is used to control the lock holding time, `RetryPrivilegeCheckProcessor` is used to extend the lock holding time regularly to prevent deadlocks and the lock is snatched)
3. High performance (there is only one piece of data in the table, the frequency of refreshing the lock ownership is not high, personally think the performance is not low, satisfied)
4. Fairness
5. Re-entrance lock (partially satisfied, since the `uniqueid` of the instance sets a guid by default at startup, considering that the id will be refreshed after restarting the process crash, so there will be a period of time when the lock cannot be re-acquired, here it should be when the process exits add the logic of releasing the lock)
In addition, regarding the mechanism of `timestamp`+`unique key`, I think my current solution is consistent. The `Id` primary key in the table is the unique key (fixed to 0), and `LeaderId` stores the instance that currently acquires the lock id, `RefreshTime` is `timestamp`, used to prevent deadlock
To sum up, I think my solution should be consistent with the implementation mechanism of this article (if I understand correctly~), but there are still some places that need further modification


@yang-xiaodong

**Comment by li-zheng-hao** [comment]
I have optimized the code structure and logic

**Comment by yang-xiaodong** [comment]
我也用中文回答了，整体改动还是不错的，一些命名可能需要调整下。

我大概看了下代码，可能对于目前来说实现过于复杂，代码量太多需要简化。

我说下我的思路（伪代码）：

1、定义  cap.lock 表

key varchar(128) not null  primary key,  -- 要锁定的对象key，如 publish_retry 或 received_retry
instance varchar(256) not null,  -- 持有锁的实例唯一Id(程序集名称+雪花Id[系统中雪花Id已考虑Mac地址，所以不会重复])
lastLockTime datetime not null  -- 持有锁的时间
added datetime not null,  -- 添加时间

2、在cap启动时除了创建表 cap.lock 结构外，还会尝试获取锁

在 IStorageInitializer 的 InitializeAsync 中
    insert cap.lock(key,instance,lastLockTime) values('publish_retry','xxxxx-xxxxx',${now}) on duplicate key ignore;
    insert cap.lock(key,instance,lastLockTime) values('received_retry','xxxxx-xxxxx',${now}) on duplicate key ignore;

成功会同时获取 publish_retry 和 received_retry 2把锁, 任何一个插入失败表示锁已经被持有，忽略。在执行sql时，不需要开事务。

3、在 MessageNeedToRetryProcessor 中，会执行一下逻辑：

在 ProcessPublishedAsync 和 ProcessReceivedAsync 去各自获取锁，这样不同实例可执行不同的任务。

获取锁的过程，update成功代表获取到锁（应该无需开启事务?）
update cap.lock set key='publish_retry',instance=${instance},lastLockTime=${now}
   WHERE key='publish_retry' and lastLockTime < ${now} - ${TTL};

在发布重试操作完成后，进行锁释放
update cap.lock set key='publish_retry',instance='',lastLockTime=${datetime.min}
   WHERE key='publish_retry' and instance=${instance};

received_retry 同理。

如果持有锁的过程中，应用崩溃，那么其他实例最多等待${TTL}可再次获取到锁。

4、关于TTL时间到达，而重试工作未完成的解决，可以在MessageNeedToRetryProcessor定义2个task.run的引用，通过判断引用是否完成来执行是否延长TTL
参考如下
```cs
var task1 = Task.Run(() =>
{
	//重试的任务
	Thread.Sleep(4000);
});

while (true)
{
	Thread.Sleep(1000);  // await context.WaitAsync(_waitingInterval)
	if(!task1.IsCompleted){
		Console.WriteLine("Task is not complate");
		//任务未完成，延长锁时间
		//update lock setlastLockTime=setlastLockTime+${TTL} where key='publish_retry' and instance=${instance}
		continue;
	}
	Console.WriteLine("Task Complate");
	break;
}
```

**Comment by li-zheng-hao** [comment]
十分感谢你的回复和指导！ 关于第三点锁的获取这一块确实是我这块经验不足想复杂了，目测了一下`update cap.lock set key='publish_retry',instance=${instance},lastLockTime=${now}
WHERE key='publish_retry' and lastLockTime < {TTL};`这个写法应该是可行的，明天有空我再看看修改一下

**Comment by li-zheng-hao** [comment]
代码已修改，关于第四点锁持有时间的延长，我没有选择在`MessageNeedToRetryProcessor`处理，我的想法是对于`MessageNeedToRetryProcessor`来说，应该只负责获取锁和释放锁，关于锁是否保持，应该由`xxxDataStorage`管理

@yang-xiaodong 对于这点请问你有什么看法吗？同时关于整体修改内容是否有进一步的建议或意见？

**Comment by yang-xiaodong** [comment]
1、我还是建议在 MessageNeedToRetryProcessor 中实现，我认为是MessageNeedToRetryProcessor需要的锁，所以应该是锁的所有者负责对锁的管理。
2、在这个应用中，不需要使用Timer实现，Timer会增加系统负载而且会有重入问题，目前1秒的间隔实例多了也会增加数据库负载。
3、我建议 TTL 设置为 options.Value.FailedRetryInterval的值 + 10秒，也就是默认70秒，在  MessageNeedToRetryProcessor中维护，通过AcquireLockAsync的参数传出去获取锁
4、在 MessageNeedToRetryProcessor 的 ProcessAsync 中，我认为只需要考虑消费侧的TTL时间到但任务未执行完成的情况，发送侧缓冲区很大，无需考虑。
5、测试csproj项目可删除，集成测试DotNetCore.CAP.MySql.Test中做，单元测试在DotNetCore.CAP.Test中，不想添加测试工程了。
6、MessageNeedToRetryProcessor的伪代码

```cs
class MessageNeedToRetryProcessor
{
	private Task _failedRetryConsumeTask;

	public async Task ProcessAsync(){
		
		//var storage = context.Provider.GetRequiredService<IDataStorage>();
		
		_ = Task.Run(ProcessPublishedAsync); //我认为只需要考虑消费侧的TTL时间到但任务未执行完成的情况，发送侧缓冲区很大，无需考虑。
		
		if(_failedRetryConsumeTask!=null && !_failedRetryConsumeTask.IsCompleted){
			// 说明 delay 60秒，消费任务未完成，那么再延长锁 60 秒
			// storage.RenewLock	
			// update lock set lastLockTime = lastLockTime + _waitingInterval 
			return;
		}

		_failedRetryConsumeTask = Task.Run(ProcessReceivedAsync);
		
		_failedRetryConsumeTask.ContinueWith(x => { 
			_failedRetryConsumeTask =null;
		});
	
		//下方会  delay 60秒
		//await context.WaitAsync(_waitingInterval).ConfigureAwait(false);  //delay 60 sec default
		
	}

	private async Task ProcessPublishedAsync()
	{
		//if (!storage.AcquireLockAsync) {
		//	return;
		//}
		
		// logic
		
		//storage.ReleaseLockAsync
	}

	private async Task ProcessReceivedAsync()
	{
		//if (!storage.AcquireLockAsync) {
		//	return;
		//}

		// logic

		//storage.ReleaseLockAsync
	}
}
```

**Comment by li-zheng-hao** [comment]
1. Renew只会在获取锁的那个实例上执行，每秒1次update的数据库负载我认为可以忽略不计？不过Timer的重入问题确实我没考虑到，这块还是按照你说的改合适
2. 集成测试的代码我下次修改直接移除，只在我本地测试

**Comment by li-zheng-hao** [comment]
已按照之前讨论修改

File: src/DotNetCore.CAP.MySql/IStorageInitializer.MySql.cs [hunk 1]
```diff
@@ -97,6 +104,16 @@ protected virtual string CreateDbTablesScript()
   PRIMARY KEY (`Id`),
   INDEX `IX_ExpiresAt`(`ExpiresAt`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+
+CREATE TABLE IF NOT EXISTS `{GetLockTableName()}` (
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP.MySql/IStorageInitializer.MySql.cs:108]: 需添加对 `_options.Value.IsUseStorageLock`的判断

File: src/DotNetCore.CAP.PostgreSql/IDataStorage.PostgreSql.cs [hunk 2]
```diff
@@ -38,6 +39,35 @@ public class PostgreSqlDataStorage : IDataStorage
...
+            var connection = new NpgsqlConnection(_options.Value.ConnectionString);
+            await using var _ = connection.ConfigureAwait(false);
+            var opResult=await connection.ExecuteNonQueryAsync(sql).ConfigureAwait(false);
+            return opResult > 0;
+        }
+
+        public async Task ReleaseLockAsync(string key,string instance, CancellationToken token = default)
+        {
+            string sql =
+                $"update {_retryName} set \"Instance\"='',\"LastLockTime\"='{DateTime.MinValue}' WHERE \"Key\"='{key}' and \"Instance\"='{instance}';";
```
Comment (yang-xiaodong) [inline src/DotNetCore.CAP.PostgreSql/IDataStorage.PostgreSql.cs:58]: SQL关键字统一使用大写
update --> UPDATE
and --> AND

下同

**Review by yang-xiaodong — CHANGES_REQUESTED** [review]
看起来不错，这2个问题处理一下我认为就可以Merge了。

File: src/DotNetCore.CAP.MySql/IStorageInitializer.MySql.cs (hunk 1 above)
Comment (li-zheng-hao) [reply src/DotNetCore.CAP.MySql/IStorageInitializer.MySql.cs:113]: 已修改

File: src/DotNetCore.CAP.PostgreSql/IDataStorage.PostgreSql.cs (hunk 2 above)
Comment (li-zheng-hao) [reply src/DotNetCore.CAP.PostgreSql/IDataStorage.PostgreSql.cs]: 已修改

**Review by yang-xiaodong — APPROVED** [review]
LGTM , Thanks for your PR!
